### PR TITLE
Local/Google Login, persisting Sessions

### PR DESCRIPTION
**Pull request by annelau99**

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -4,30 +4,50 @@ const cors = require('cors');
...
 
 const startDB = (done) => {
   models.init((database) => {
     userRoutes.init(database);
     groupRoutes.init(database);
+    passportSetup.init(database);
     db = database;
-    db.sequelize.sync({ force: true }).then(() => {
+    db.sequelize.sync({force: true}).then(() => {
+      db.sequelize.sync;
```
Comment (StevieSong) [inline server/index.js]: delete this line

File: server/routes/auth-routes.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+const router = require('express').Router();
+const passport = require('passport');
+let client_add = "http://localhost:3000";
+
+// function ensureAuthenticated
+// ensures the user is logged in before it grants access to the api.
+// input: req, the request; res, the response obj; next, the function to be called next
+function ensureAuthenticated(req, res, next)
+{
```
Comment (StevieSong) [inline server/routes/auth-routes.js]: curly brace on same line

File: server/routes/auth-routes.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    if (req.isAuthenticated()) {
+        return next();
+    } else {
+        res.status(403).json( {"failure": "you are not logged in"});
```
Comment (StevieSong) [inline server/routes/auth-routes.js]: use 401 if not logged in. 403 is if you are authenticated but forbidden from accessing the resource. 
rename failure to err for consistency

File: server/routes/auth-routes.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+router.get('/logout', ensureAuthenticated, (req,res) => {
+    req.logout();
+    // TODO: detect whether logout is successful
+    res.send('Success');
+  // handle with passport
+});
+
+// auth with our own Login
+router.get('/local', (req, res) => {
+    res.sent("Error: GET method not supported.");
```
Comment (StevieSong) [inline server/routes/auth-routes.js]: what's the point of this route?

File: server/config/config.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "development": {
-    "username": "me",
+    "username": "peterlam",
```
Comment (StevieSong) [inline server/config/config.json]: change back to me

File: server/routes/auth-routes.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            "email": req.user.email,
+            "firstName": req.user.firstName,
+            "lastName": req.user.lastName
+        }
+    });
+});
+
+// auth with our own signup
+router.get('/signup', (req, res) => {
+  res.render('signup');
```
Comment (StevieSong) [inline server/routes/auth-routes.js]: why are we rendering views with non-react?

File: server/routes/auth-routes.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  // handle with passport
+});
+
+// auth with our own Login
+router.get('/local', (req, res) => {
+    res.sent("Error: GET method not supported.");
+});
+
+router.post('/local', passport.authenticate('local'), (req, res) => {
+    res.json({
```
Comment (StevieSong) [inline server/routes/auth-routes.js]: redirect to route handler for GET user, there's alot of logic there that this doesn't encompass

File: server/routes/auth-routes.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  passport.authenticate('google', {"failureRedirect": "/google"}),
+  (req,res) => {
+      var name = req.user.firstName + " " + req.user.lastName;
+      res.send("<html><p>you reached the callback URI. </p><button onclick='window.close()'>Close Window</button>" +
+          "<script>window.onload = () => {console.log(\"sending msg\");" +
+          "window.opener.postMessage(\"" + name + "\", \"" + client_add + "\");window.close()};</script></html>");
+  }
+);
+
+router.get('/loginstatus', (req, res) => {
```
Comment (StevieSong) [inline server/routes/auth-routes.js]: use case for this route?

File: server/views/app.ejs [hunk 9]
```diff
@@ -0,0 +1,55 @@
+<!DOCTYPE html>
```
Comment (StevieSong) [inline server/views/app.ejs]: see comment regarding use of non-react view engine, in fact do we still need these?

File: server/routes/userRoutes.js [hunk 10]
```diff
@@ -37,6 +37,9 @@ router.post('/create', (req, res) => {
     password: req.body.password // TODO hash password
   };
 
+  console.log(req.body);
```
Comment (StevieSong) [inline server/routes/userRoutes.js]: delete console logs

File: client/src/Authentication/LoginForm.js [hunk 11]
```diff
@@ -1,86 +1,208 @@
...
+            },
+            credentials: 'include',
+            body: JSON.stringify(userInfo)
+        });
+
+        const myJson = await response.json();
+        console.log(JSON.stringify(myJson));
+
+        if (response.status === 200) {
+            alert("registration successful");
```
Comment (StevieSong) [inline client/src/Authentication/LoginForm.js]: should redirect

File: server/index.js [hunk 12]
```diff
@@ -4,30 +4,50 @@ const cors = require('cors');
...
+app.use(cors(corsOptions));
+app.use(require('cookie-parser')());
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: true }));
 
+app.use(session({secret: 'some_secret_key', resave: false, saveUninitialized: false }));
+app.use(passport.initialize());
+app.use(passport.session());
+
+app.use('/auth', authRoutes);
```
Comment (StevieSong) [inline server/index.js]: can we prepend this with /api as well

Comment (waymao) [reply server/index.js]: @harshyyy21 will need to update his google API settings but yes.

File: server/routes/auth-routes.js (hunk 8 above)
Comment (waymao) [reply server/routes/auth-routes.js]: to get info for the user currently logged in

**Comment by waymao** [comment]
persist login has now been implemented (by changing the cookie expiration time to 1 year). Maybe we need to add a check box to ask user to "keep me logged in"? (that should add some extra work to the backend and we'll see)

**Comment by waymao** [comment]
most change requests have also been implemented

File: server/index.js (hunk 12 above)
Comment (waymao) [reply server/index.js]: updated

**Pull request merged**